src/hbbft_rbc.erl

```diff
@@ -170,8 +217,46 @@ init_test() ->
     {_, ConvergedResults} = hbbft_test_utils:do_send_outer(?MODULE, [{0, {send, MsgsToSend}}], StatesWithId, sets:new()),
     %% everyone should converge
     ?assertEqual(N, sets:size(ConvergedResults)),
+    %% the decoded result should be the original message
+    ConvergedResultsList = sets:to_list(ConvergedResults),
+    ?assert(lists:all(fun({result, {_, Res}}) -> Res == Msg end, ConvergedResultsList)),
     ok.
 
+send_incorrect_msg_test() ->
+    N = 5,
+    F = 1,
+    Msg = crypto:strong_rand_bytes(512),
+    S0 = hbbft_rbc:init(N, F),
+    S1 = hbbft_rbc:init(N, F),
+    S2 = hbbft_rbc:init(N, F),
+    S3 = hbbft_rbc:init(N, F),
+    S4 = hbbft_rbc:init(N, F),
+    {NewS0, {send, MsgsToSend}} = hbbft_rbc:input(S0, Msg),
+
+    %% ====================================================
+    %% screw up 3 val messages in the MsgsToSend
+    %% TODO: something better but this works for now
+    BadMsg = crypto:strong_rand_bytes(512),
+    M = N - 2*F,
+    K = 2*F,
+    {ok, Shards} = leo_erasure:encode({K, M}, BadMsg),
+    MsgSize = byte_size(BadMsg),
+    ShardsWithSize = [{MsgSize, Shard} || Shard <- Shards],
+    Merkle = merkerl:new(ShardsWithSize, fun merkerl:hash_value/1),
+    MerkleRootHash = merkerl:root_hash(Merkle),
+    BranchesForShards = [merkerl:gen_proof(Hash, Merkle) || {Hash, _} <- merkerl:leaves(Merkle)],
+    BadMsgsToSend = [ {unicast, J, {val, MerkleRootHash, lists:nth(J+1, BranchesForShards), lists:nth(J+1, ShardsWithSize)}} || J <- lists:seq(0, N-1)],
+    [ First, Second | _ ] = MsgsToSend,
+    [ Third, Fourth, Fifth | _ ] = BadMsgsToSend,
```

Comment:
You want to change this to `[ _, _, Third, Fourth, Fifth] = BadMsgsToSend` and then 4 of the 5 nodes will converge on the aborted result.

The node with the original data input doesn't converge, annoyingly, because I think it gets enough agreeing shards that it proceeds onto sending readies.

Reply:
I think a larger issue here is the `h` in the state vs the `h` in the message, what should happen if they disagree (we simply clobber the old value right now).

Additionally, we should have a test where we inject VAL messages with invalid proofs (strangely we don't check the merkle proof is valid on the VAL path).

Reply:
Reading RBC from the paper more closely, I think we need to group shares, num_readies, num_echoes, seen_val and ready_sent by `h` and check the conditions for each seen value of `h`?

Reply:
Right, I have to spend some time on having a new test which triggers the `aborted` path, I've corrected the test as per your suggestion for now.

As for _we don't check the merkle proof is valid on the VAL path_, I don't think that's mentioned in the whitepaper RBC protocol either. You receive a VAL msg you multicast ECHO and that's it. Perhaps a valid question for professor Miller?